api/src/routes/protected/user.ts

```diff
@@ -114,6 +114,58 @@ export const userRoutes: FastifyPluginCallbackTypebox = (
     }
   );
 
+  fastify.delete(
+    '/users/:userId',
+    {
+      schema: schemas.deleteMyAccount
+    },
+    async (req, reply) => {
+      const logger = fastify.log.child({ req, res: reply });
+      const { userId } = req.params as { userId: string };
+
+      if (userId !== req.user?.id) {
+        logger.warn(
+          { requestedUserId: userId, authUserId: req.user?.id },
+          'User attempted to delete a different account.'
+        );
+        void reply.code(403);
+        return { type: 'error', message: 'forbidden' } as const;
+      }
+
+      logger.info(`User ${req.user?.id} requested account deletion`);
+      await fastify.prisma.userToken.deleteMany({
+        where: { userId: req.user.id }
+      });
+      await fastify.prisma.msUsername.deleteMany({
+        where: { userId: req.user.id }
+      });
+      await fastify.prisma.survey.deleteMany({
+        where: { userId: req.user.id }
+      });
+      try {
+        await fastify.prisma.user.delete({
+          where: { id: req.user.id }
+        });
+      } catch (err) {
+        if (
+          err instanceof PrismaClientKnownRequestError &&
+          err.code === 'P2025'
+        ) {
+          logger.warn(
+            err,
+            `User with id ${req.user?.id} not found for deletion.`
+          );
+        } else {
+          logger.error(err, 'Error deleting user account');
+          throw err;
+        }
+      }
+      reply.clearOurCookies();
+
+      return {};
```

Comment:
It is common for `DELETE`s to return no body with a status of `204`:
> https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods/DELETE#successfully_deleting_a_resource

```suggestion
      return reply.code(204);
```